easypermissions/src/main/java/pub/devrel/easypermissions/EasyPermissions.java

```diff
@@ -314,92 +324,187 @@ public static void onRequestPermissionsResult(int requestCode, @NonNull String[]
+     * @param deniedPermission denied permission.
+     * @return {@code true} if the permissions has been permanently denied.
+     */
+    public static boolean permissionPermanentlyDenied(@NonNull Activity activity,
+                                                      @NonNull String deniedPermission) {
+        return !shouldShowRequestPermissionRationale(activity, deniedPermission);
+    }
+
+    /**
+     * Check if a permission has been permanently denied (user clicked "Never ask again").
+     *
+     * @param fragment         Fragment requesting permissions.
+     * @param deniedPermission denied permission.
+     * @return {@code true} if the permissions has been permanently denied.
+     */
+    public static boolean permissionPermanentlyDenied(@NonNull Fragment fragment,
+                                                      @NonNull String deniedPermission) {
+        return !shouldShowRequestPermissionRationale(fragment, deniedPermission);
+    }
+
+    /**
+     * Check if a permission has been permanently denied (user clicked "Never ask again").
+     *
+     * @param fragment         android.app.Fragment requesting permissions.
+     * @param deniedPermission denied permission.
+     * @return {@code true} if the permissions has been permanently denied.
+     */
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    public static boolean permissionPermanentlyDenied(@NonNull android.app.Fragment fragment,
+                                                      @NonNull String deniedPermission) {
+        return !shouldShowRequestPermissionRationale(fragment, deniedPermission);
+    }
+
+    private static void notifyAlreadyHasPermissions(Object object,
+                                                    int requestCode,
+                                                    @NonNull String[] perms) {
+        int[] grantResults = new int[perms.length];
+        for (int i = 0; i < perms.length; i++) {
+            grantResults[i] = PackageManager.PERMISSION_GRANTED;
         }
 
-        return null;
+        onRequestPermissionsResult(requestCode, perms, grantResults, object);
     }
 
-    @Nullable
-    private static android.app.FragmentManager getFragmentManager(@NonNull Object object) {
+
+    /**
+     * @param object Activity or Fragment
+     * @return true if the user has previously denied any of the {@code perms} and we should show a
+     * rationale, false otherwise.
+     */
+    private static boolean shouldShowRationale(@NonNull Object object, @NonNull String[] perms) {
+        boolean shouldShowRationale = false;
+        for (String perm : perms) {
+            shouldShowRationale =
+                    shouldShowRationale || shouldShowRequestPermissionRationale(object, perm);
+        }
+        return shouldShowRationale;
+    }
+
+    private static boolean shouldShowRequestPermissionRationale(@NonNull Object object,
+                                                                @NonNull String perm) {
```

Comment:
This is the only one I left with `Object` because it helps reduce triplicating logic. (I didn't know `triplicating` was a word, lol 😄)